zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/CassandraZipkinQueryDAO.java

```diff
@@ -0,0 +1,297 @@
+      if (StringUtil.isNotEmpty(request.serviceName())) {
+        completionTraceIds.add(client.executeAsyncQuery("select " + ZipkinSpanRecord.TRACE_ID + " from " + table +
+                " where " + ZipkinSpanRecord.LOCAL_ENDPOINT_SERVICE_NAME + " = ?" +
+                " and " + ZipkinSpanRecord.TIME_BUCKET + " >= ?" +
+                " and " + ZipkinSpanRecord.TIME_BUCKET + " <= ? ALLOW FILTERING",
+            row -> row.getString(ZipkinSpanRecord.TRACE_ID),
+            request.serviceName(), duration.getStartTimeBucket(), duration.getEndTimeBucket()
+        ));
+      }
+      if (StringUtil.isNotEmpty(request.remoteServiceName())) {
+        completionTraceIds.add(client.executeAsyncQuery("select " + ZipkinSpanRecord.TRACE_ID + " from " + table +
+                " where " + ZipkinSpanRecord.REMOTE_ENDPOINT_SERVICE_NAME + " = ?" +
+                " and " + ZipkinSpanRecord.TIME_BUCKET + " >= ?" +
+                " and " + ZipkinSpanRecord.TIME_BUCKET + " <= ? ALLOW FILTERING",
+            row -> row.getString(ZipkinSpanRecord.TRACE_ID),
+            request.remoteServiceName(), duration.getStartTimeBucket(), duration.getEndTimeBucket()
+        ));
+      }
+      if (StringUtil.isNotEmpty(request.spanName())) {
+        completionTraceIds.add(client.executeAsyncQuery("select " + ZipkinSpanRecord.TRACE_ID + " from " + table +
+                " where " + ZipkinSpanRecord.NAME + " = ?" +
+                " and " + ZipkinSpanRecord.TIME_BUCKET + " >= ?" +
+                " and " + ZipkinSpanRecord.TIME_BUCKET + " <= ? ALLOW FILTERING",
+            row -> row.getString(ZipkinSpanRecord.TRACE_ID),
+            request.spanName(), duration.getStartTimeBucket(), duration.getEndTimeBucket()
+        ));
+      }
+
+      traceIdSet.addAll(retainTraceIdList(completionTraceIds));
+    }
+
+    return getTraces(request.limit() > 0 ? traceIdSet.stream().limit(request.limit()).collect(Collectors.toSet()) : traceIdSet);
+  }
+
+  private Set<String> retainTraceIdList(List<CompletionStage<List<String>>> completionStages) {
+    Iterator<CompletionStage<List<String>>> iterator = completionStages.iterator();
+    if (!iterator.hasNext()) return Collections.emptySet();
+    Set<String> result = new HashSet<>(iterator.next().toCompletableFuture().join());
+    while (iterator.hasNext() && result.size() > 0) {
+      Set<String> nextSet = new HashSet<>(iterator.next().toCompletableFuture().join());
+      result.retainAll(nextSet);
+    }
+
+    return result;
+  }
+
+  @Override
+  public List<List<Span>> getTraces(Set<String> traceIds) throws IOException {
+    if (CollectionUtils.isEmpty(traceIds)) {
+      return Collections.emptyList();
+    }
+
+    final List<List<Span>> result = new ArrayList<>();
+    for (String table : tableHelper.getTablesWithinTTL(ZipkinSpanRecord.INDEX_NAME)) {
+      final PreparedStatement stmt = client.getSession().prepare("select * from " + table + " where " +
+          ZipkinSpanRecord.TRACE_ID + " in ?");
```

Comment:
don't use `IN ` against. the primary key. async concurrent requests instead.

Reply:
Updated to using multiple equals query.